fix bugs in quickstart docs

adamfeuer opened the pull request:
## Summary

- fix bugs in quickstart docs (see #1720)
  - change Gnu to GNU
  - fix formatting and bad links
  - add menuconfig screenshot
  - replaced config file edits with kconfig-tweak
  - fix section numbering
  - add `custinfo()` function to debug.h and Kconfig for custom logging facility

## Impact

- should fix all doc issues listed in #1720 

## Testing

- built manually locally

v01d commented on Documentation/quickstart/debugging.rst:
```diff
@@ -93,11 +98,11 @@ You can add your own custom debug logging by adding the following lines to
-You need to add the following line to your ``.config`` file:
+You need to enable the setting this way:
 
 .. code-block:: c
```
```suggestion
.. code-block:: console
```

v01d commented on Documentation/quickstart/debugging.rst:
```diff
@@ -49,30 +54,30 @@ Changing Debug Settings Quickly
-   $ kconfig-tweak --disable CONFIG_DEBUG_SYMBOLS
-   $ kconfig-tweak --disable CONFIG_DEBUG_NOOPT
-   $ kconfig-tweak --disable CONFIG_SYSLOG_TIMESTAMP
+   kconfig-tweak --disable CONFIG_DEBUG_ALERT
```
I think for `console` you need the prompt. Otherwise it is assumed to be output of a command and it appears dimmer.

adamfeuer replied:
@v01d Ok, thanks for catching. I accepted your changes.

v01d commented on Documentation/quickstart/install.rst:
```diff
@@ -16,24 +16,23 @@ Install prerequisites for building using Linux
    This is done by adding your Linux user to the ``dialout`` group:
 
- .. code-block:: bash
+    .. code-block:: bash
```
```suggestion
    .. code-block:: console
```

v01d commented on Documentation/quickstart/install.rst:
```diff
@@ -82,48 +81,48 @@ There are a collection of required tools that need to be built to build most Apa
+   This required to build ``kconfig-frontends``.
 
- .. code-block:: bash
+   .. code-block:: bash
```
```suggestion
   .. code-block:: console
```

v01d commented on Documentation/quickstart/install.rst:
```diff
@@ -82,48 +81,48 @@ There are a collection of required tools that need to be built to build most Apa
+   This is necessary to run the ``./nuttx/tools/configure.sh`` script as well as using the ncurses-based menuconfig system.
 
- .. code-block:: bash
+   .. code-block:: bash
```
```suggestion
   .. code-block:: console
```

v01d commented on Documentation/quickstart/install.rst:
```diff
@@ -82,48 +81,48 @@ There are a collection of required tools that need to be built to build most Apa
    This is required if you want to build ROM-based file systems.
 
- .. code-block:: bash
+   .. code-block:: bash
```
```suggestion
   .. code-block:: console
```

v01d commented on Documentation/quickstart/install.rst:
```diff
@@ -138,20 +137,20 @@ to use it, modify it or help develop it, you'll need the source code.
 You can either clone the public repositories:
 
-    .. code-block:: bash
+.. code-block:: bash
```
```suggestion
.. code-block:: console
```

v01d commented on Documentation/quickstart/install.rst:
```diff
@@ -138,20 +137,20 @@ to use it, modify it or help develop it, you'll need the source code.
 Or, download the `tarball <https://github.com/apache/incubator-nuttx/tarball/master>`_:
 
-    .. code-block:: bash
+.. code-block:: bash
```
```suggestion
.. code-block:: console
```

v01d commented on Documentation/quickstart/install.rst:
```diff
@@ -168,27 +167,27 @@ binary file on your embedded computer. This guide assumes your computer is an
-    $ # add the toolchain bin/ dir to your path...
-    $ # you can edit your shell's rc files if you don't use bash
-    $ echo "export PATH=/opt/gcc/gcc-arm-none-eabi-9-2019-q4-major/bin:$PATH" >> ~/.bashrc
+.. code-block:: bash
```
```suggestion
.. code-block:: console
```

v01d commented:
There's a leftover todo on quickstart/index.rst
Otherwise this looks LGTM.

v01d commented:
If you don't mind, please squash once you are ready.

adamfeuer commented:
@v01d Thanks for catching all those issues. I rebased and squashed and force-pushed. If @btashton is good with the fixes, it should be good to merge.

BTW we should add interactive rebase to the docs to show how to squash the commits. 🙂

btashton commented on Documentation/quickstart/debugging.rst:
```diff
@@ -6,35 +6,40 @@ Debugging
+   $ kconfig-tweak --enable CONFIG_SYSLOG_TIMESTAMP
 
 You may need to do a little bit of experimenting to find the combination of logging settings that work for the problem
 you're trying to solve. See the file `debug.h <https://github.com/starcat-io/incubator-nuttx/blob/master/include/debug.h>`_
```
Did we intend to link to your fork?

adamfeuer replied:
@btashton No, that's an error. Will fix.

btashton commented on Documentation/quickstart/debugging.rst:
```diff
@@ -93,11 +98,11 @@ You can add your own custom debug logging by adding the following lines to
+.. code-block:: console
 
-   CONFIG_DEBUG_CUSTOM_INFO=y
+   $ kconfig-tweak --enable CONFIG_CUSTOM_INFO
```
I dont think this will work because there is no Kconfig that defines CUSTOM_INFO.  Might be best in the block above to not depend on CONFIG_DEBUG_CUSTOM_INFO and just do the `#define custinfo  _info`.  Or recommend just using `_info` for debugging.

adamfeuer replied:
@btashton using _info for logging produces too much output... How about I add a kconfig entry for CUSTOM_INFO? I could also add the code to debug.h too. That would save people from having to add it.

It's super handy to have a logging statement that you can use for your own purposes that no other module is using.

btashton replied:
I would be open to that with the clear expectation that we would not allow usage of it to be merged in.

btashton replied:
also if you do add that, the `custinfo` should be defined in the header file

adamfeuer replied:
@btashton Ok, I added custinfo to debug.h, CUSTOM_INFO to the Kconfig file, and updated docs.

I see what you mean about using the Kconfig menu system to set this– testing it, I can see how it work. Some good instructions would be helpful as an alternative to the kconfig-tweak stuff... but it really needs an asciinema movie to show people what to do... maybe in another PR??

adamfeuer replied:
@btashton I added a comment in debug.h that says "do not create PRs that include usages of custinfo" – let me know if you want a different message there.

Ouss4 replied:
Where is this going to be used?  If that's something just to track an issue, why not just `printf` that we delete afterwards?
If this is part of a driver, I think using what's available* is readable enough. 

*: I mean ainfo for ADC, lcdinfo for LCDs, etc.

adamfeuer replied:
@Ouss4 I didn't find `printf` was enough– I wanted the timestamps that the logging facility produces, so would have had to reinvent that myself if I used `printf` – and `printf` also cannot use the in-memory logging.

While porting the imxrt SDMMC driver to sama5, I found it very helpful to have an independent logging facility that I could use to track timing issues across multiple modules and multiple threads, without unleashing 🙂 a torrent of logging from other modules.

btashton replied:
@adamfeuer this is where I am also confused because developing that driver I would expect you would have used `mcinfo` which is controlled by `CONFIG_DEBUG_MEMCARD_INFO`  and Kconfig describes as "Enable MMC/SD memory card driver informational output to SYSLOG."

example:  `arch/arm/src/stm32/stm32_sdio.c`

I guess I would just expect people to use the subsystem logger they are building on or add one if there is not an appropriate.

btashton replied:
I am going AFK for the rest of the holiday, so don't let me hold this up :)

adamfeuer replied:
@btashton The issue with solely using `mcinfo` is that I had to look into several other different subsystems – `fs` which uses `fsinfo`, and a couple more that I don't remember right now. With a cross-cutting concern, if you turn on all the logging for several subsytems, that unleashes a torrent of logging. I needed something that I could add to different subsystems without turning everything on.

Anyway, if this is too specialized, I can either go back to having it be a tip in the docs, or remove it completely from the docs and just document it on the wiki or my own website.

Have a great holiday!

adamfeuer replied:
@Ouss4 @btashton I removed the custinfo Kconfig, debug.h, and docs so that the rest of the error fixes can go in.

btashton commented on Documentation/quickstart/debugging.rst:
```diff
@@ -6,35 +6,40 @@ Debugging
+   kconfig-tweak --enable CONFIG_DEBUG_SYMBOLS
+   kconfig-tweak --enable CONFIG_DEBUG_NOOPT
+
+Note that enabling all these will produce an incredible amount of logging output. Enable the level you want and
```
Is there a reason we are not recommending that people explore the menuconfig?  I get that `kconfig-tweak` is useful for turning things on and off, but searching through header files seems to be working against the while point of menuconfig.

adamfeuer replied:
@btashton Well, when I wrote this, I found using the menuconfig system to turn on and off logging levels to be incredibly frustrating. I found navigating and finding things very difficult. So I resorted to this system. 

If you have another way to easily turn on and off logging, or we can refactor the menuconfig system to make this easier, I'm in. Let me know.

btashton replied:
I just turn on all the logging levels and then enable logging level for whatever system I need.  The only gotcha here is that you need to turn on the global levels first.  I also keep that enabled in my defconfig along with things like debug symbols until I am ready to upstream.

My biggest objection is that using things like kconfig-tweak and searching header files is that people do not resolve dependencies correctly or find useful features for what they have enabled.  Plus you have docs in menuconfig.

btashton replied:
I also would like to add some https://asciinema.org/ demos which we can embed in the docs, but that is for later.
https://pypi.org/project/sphinxcontrib.asciinema/

adamfeuer replied:
@btashton If you have a vision for how to rewrite this section, I would love to see that! I would love to learn a better way. I am a NuttX newbie. 

I wrote this after being frustrated from a series of long hard driver debugging sessions, from the perspective of someone who knew nothing about NuttX internals and had no guidance about how to learn what was happening inside NuttX. This was my best attempt at telling others information that I wanted to have...

adamfeuer replied:
@btashton The asciinema stuff sounds great too.

btashton replied:
@adamfeuer Sounds good, we can leave this as is lets just make sure we address the custinfo thing.

adamfeuer commented:
@v01d I removed the custinfo Kconfig, debug.h, and docs since it seems controversial, maybe I'll submit it in another PR. @btashton was ok with the rest of the fixes.

Would you be willing to merge this?

Ouss4 commented:
I have to revive the kconfig-tweek vs menuconfig discussion.

I find debug output the easiest of all CONFIGs_.  They are just one after the other.  Of course you have to first enable a subsystem before you can see its debug options (CONFIG_NET for CONFIG_DEBUG_NET, CONFIG_CRYPTO for CONFIG_DEBUG_CRYPTO, etc.).
Besides enabling them through kconfig-tweek will just write to the .config file, you'd still have to do `olddefconfig` (or any other config variant including `menuconfig`) to correctly update the .config file.
I don't see that any easier than searching with menuconfig.  BTW, one can search in menuconfig with "/".

adamfeuer commented:
@Ouss4 Do you want me to rewrite this instruction using the menuconfig system? Or add instructions for menuconfig too? Or...?

Ouss4 commented:
>  @Ouss4 Do you want me to rewrite this instruction using the menuconfig system?

I definitely prefer that.  What do you all think?  Please let know how I can help.

v01d commented:
Yes, I think it is better. I would consider `kconfig-tweak` use a more "advanced" topic.

adamfeuer commented:
> I definitely prefer that. What do you all think? Please let know how I can help.

Cool. 🙂 I will try now, I'll push changes here– it would be a great help to review them once I get it done.

adamfeuer commented:
@Ouss4 Ok, I rewrote the debugging section to use the `menuconfig`, and added a screenshot — will you check it out and see what you think? If you have changes or improvements I would love to know them.

adamfeuer commented:
@Ouss4 In the diff view there is a RST rich diff button available on the upper right in the files view, by the way, that will show the rendered RST (minus screenshot).

Ouss4 commented on Documentation/quickstart/debugging.rst:
```diff
@@ -6,38 +6,49 @@ Debugging
+You can then select from the subsystems that are available, Network, Scheduler, USB, etc. Note that you will need to
+separately enable the subsystem elsewhere in the ``menuconfig`` system. To see the ``CONFIG`` define that is set,
+use the arrow keys to highlight the subsystem (for instance, ``Network Debug Features``) and type '?'. This will show
+you that the C constant that is set is called ``CONFIG_DEBUG_NET``. ``debug.h`` defines the ``netinfo()`` logging
```
```suggestion
you that the C macro that is set is called ``CONFIG_DEBUG_NET``. ``debug.h`` defines the ``netinfo()`` logging
```

Ouss4 commented on Documentation/quickstart/debugging.rst:
```diff
@@ -6,38 +6,49 @@ Debugging
+separately enable the subsystem elsewhere in the ``menuconfig`` system. To see the ``CONFIG`` define that is set,
+use the arrow keys to highlight the subsystem (for instance, ``Network Debug Features``) and type '?'. This will show
+you that the C constant that is set is called ``CONFIG_DEBUG_NET``. ``debug.h`` defines the ``netinfo()`` logging
+function that will log output if this constant is set. You can search the source code for ``netinfo`` to see how it is
```
```suggestion
function that will log output if this macro is set. You can search the source code for ``netinfo`` to see how it is
```

Ouss4 commented on Documentation/quickstart/debugging.rst:
```diff
@@ -49,63 +60,33 @@ Changing Debug Settings Quickly
+.. code-block:: console
 
    $ kconfig-tweak --disable CONFIG_DEBUG_NET
+   $ make oldconfig  # needed to have the kconfig system check the config
```
I find olddefconfig better as it will without prompting set new symbols to their default value.

adamfeuer replied:
@Ouss4 Cool, thanks. I accepted your suggestions.

Ouss4 commented on Documentation/quickstart/debugging.rst:
```diff
@@ -49,63 +60,33 @@ Changing Debug Settings Quickly
    $ kconfig-tweak --disable CONFIG_DEBUG_NET
+   $ make oldconfig  # needed to have the kconfig system check the config
    $ kconfig-tweak --enable CONFIG_DEBUG_NET
+   $ make oldconfig
```
```suggestion
   $ make olddefconfig
```

Ouss4 reviewed (commented):
Thanks @adamfeuer, please consider the following:

Ouss4 commented on Documentation/quickstart/debugging.rst:
```diff
@@ -49,63 +60,33 @@ Changing Debug Settings Quickly
+.. code-block:: console
 
    $ kconfig-tweak --disable CONFIG_DEBUG_NET
+   $ make oldconfig  # needed to have the kconfig system check the config
```
```suggestion
   $ make olddefconfig  # needed to have the kconfig system check the config
```
Sorry I didn't use the suggestion with this one.

adamfeuer commented:
@Ouss4 Thanks for catching that! Ok, rebased, squashed, force-pushed, should be good to go. Do you have other improvements or comments?

Ouss4 commented:
Thanks Adam, LGTM.  If @v01d or @btashton have no comments, we can merge this.

v01d commented:
LGTM

v01d commented:
Brennan is AFK I believe, btw.

adamfeuer commented:
@Ouss4 I think @btashton is out until Tuesday morning due to a US holiday, but he ok'd everything by the `kconfig-tweak` thing, which I rewrote at your request. He and said not to hold up merging because of him ([see above](https://github.com/apache/incubator-nuttx/pull/1722#discussion_r484008244)). @v01d what do you think?

Ouss4 commented:
All right, then. Merging.. Thanks @adamfeuer.
